src/lib.rs

```diff
@@ -259,15 +259,34 @@ fn serve_request_stream<Ctx: Clone>(
 ) -> impl Stream<Item = Result<Message, warp::Error>> {
     let resp_stream = srv
         .serve_ws(ctx, payload, service_id)
-        .map(move |payload_result| match payload_result {
-            Ok(payload) => Outgoing::Next {
-                request_id: req_id,
-                payload,
-            },
-            Err(kind) => Outgoing::Error {
-                request_id: req_id,
-                kind,
-            },
+        .take_until_condition(|resp| future::ready(resp.is_err()))
+        .ready_chunks(128)
+        .flat_map(move |payload_results| {
+            let mut err = None;
+            let mut payload = Vec::with_capacity(payload_results.len());
+            for payload_result in payload_results {
+                match payload_result {
+                    Ok(value) => payload.push(value),
+                    Err(kind) => {
+                        err = Some(kind);
+                        break;
+                    }
+                }
+            }
+            let mut res = Vec::with_capacity(1);
+            if !payload.is_empty() {
+                res.push(Outgoing::Next {
+                    request_id: req_id,
+                    payload,
+                });
+            }
+            if let Some(kind) = err {
+                res.push(Outgoing::Error {
+                    request_id: req_id,
+                    kind,
+                });
+            }
+            stream::iter(res)
```

Comment:
is there a more elegant way to group ok values and potential trailing err?

Reply:
since only the last element can be Err, how about
```rust
let err = if payload_results.last().iter().any(|r| r.is_err()) {
    Some(payload_results.pop().unwrap())
} else {
    None
}
let payload = payload_results.into_iter().collect::<Result<_>>().unwrap();
...
res.extend(err.map(...));
```

Reply:
That will panic if for some reason the error is not at the end. And it is not ensured that this will be the case. So I would prefer to actually look for the first error.

Reply:
Due to the `take_until_condition` there cannot possibly be an error somewhere in the middle.

Reply:
Ah, right.

Reply:
thanks for your ideas, guys! i played with it but it doesn't get significantly better IMO. 🤷‍♂️